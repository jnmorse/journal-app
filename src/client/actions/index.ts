export * from './types';
export * from './interfaces';
export * from './user-actions';
