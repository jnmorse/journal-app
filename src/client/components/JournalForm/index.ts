import { connect } from 'react-redux';

import JournalForm from './JournalForm';

export default connect()(JournalForm);
