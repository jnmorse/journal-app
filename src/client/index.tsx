import React from 'react';
import ReactDOM from 'react-dom';
import { hot } from 'react-hot-loader/root';

import { Root } from './Root';

const HotRoot = hot(Root);

import './index.css';

ReactDOM.render(<Root />, document.getElementById('root'));
