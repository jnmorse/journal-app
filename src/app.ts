import express, { Request, Response, Application, Errback } from 'express';
import compression from 'compression';
import morgan from 'morgan';
import helmet from 'helmet';
import { NextFunction } from 'connect';
import { StaticRouterContext } from 'react-router';
import passport from 'passport';
import expressSession, { MemoryStore } from 'express-session';
import mongoDBStore from 'connect-mongodb-session';
import dotenv from 'dotenv';

import { reactRenderer } from './middleware/react-renderer';
import { userRoutes } from './api-routes/user-routes';
import { reduxStoreMiddleware } from './middleware/redux-store';
import { Reducers } from './client/reducers';
import { JournalRoutes } from './api-routes/journal-routes';

const app: Application = express();

const MongoStore = mongoDBStore(expressSession);
dotenv.config();

const notTestEnv = process.env.NODE_ENV !== 'test';
const isDevelopment = process.env.NODE_ENV === 'development';

app.use([
  express.json(),
  express.urlencoded({ extended: true }),
  expressSession({
    cookie: {
      maxAge: 1000 * 60 * 60 * 24 * 7
    },
    store: new MongoStore({
      uri: process.env.SESSION_URL || '',
      collection: 'sessions'
    }),
    secret: process.env.COOKIE_SECRET || 'MySecret',
    resave: true,
    saveUninitialized: true
  }),
  helmet(),
  compression()
]);

app.use(passport.initialize());
app.use(passport.session());

if (notTestEnv && isDevelopment) {
  app.use(morgan('dev'));
} else if (notTestEnv) {
  app.use(morgan('combined'));
}

app.get('/api/test', (req: Request, res: Response): void => {
  res.json({ message: 'Hello World' });
});

app.use('/api', userRoutes);
app.use('/api', JournalRoutes);

app.use(
  (
    error: Errback,
    req: Request,
    res: Response,
    next: NextFunction
  ): void | Response => {
    if (error) {
      return res.status(500).send(error);
    }

    return next();
  }
);

app.get('/favicon.ico', (req, res) => {
  res.status(404).send('Not Okay');
});

app.use('/css', express.static('dist/css'));
app.use('/js', express.static('dist/js'));
app.use('/images', express.static('dist/images'));
app.use('/manifest.json', express.static('dist/manifest.json'));
app.get('/robots.txt', (req, res) => {
  res.send('');
});

app.get('/favicon.ico', (req, res) => {
  res.sendStatus(404);
});

app.get(
  '*',
  reduxStoreMiddleware(Reducers),
  reactRenderer(),
  (req: Request, res: Response): Response | void => {
    const { context }: { context: StaticRouterContext } = res.locals;

    if (context.url && context.statusCode) {
      return res.redirect(context.statusCode, context.url);
    }

    if (context.statusCode) {
      res.status(context.statusCode);
    }

    const html: string = res.locals.html;

    return res.send(html);
  }
);

export { app };
