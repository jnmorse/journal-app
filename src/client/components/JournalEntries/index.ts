import { connect } from 'react-redux';

import JournalEntries from './JournalEntries';
import { StoreState } from '../../reducers';
import { JournalEntry, getJournalEntries } from '../../actions';
import { ThunkDispatch } from 'redux-thunk';
import { AxiosInstance } from 'axios';
import { Action, Dispatch } from 'redux';

export interface JournalStateProps {
  journals: JournalEntry[];
}

export interface DispatchProps {
  getJournalEntries: (limit?: number, offset?: number) => Promise<void>;
}

function mapStateToProps({ journals }: StoreState): JournalStateProps {
  return { journals };
}

function mapDispatchToProps(
  dispatch: ThunkDispatch<StoreState, AxiosInstance, Action>
): DispatchProps {
  return {
    getJournalEntries: (limit?: number, offset?: number) =>
      dispatch(getJournalEntries(limit, offset))
  };
}

export default connect<
  JournalStateProps,
  DispatchProps,
  { limit: number; offset: number },
  StoreState
>(
  mapStateToProps,
  mapDispatchToProps
)(JournalEntries);
