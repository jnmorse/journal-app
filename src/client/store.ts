import { createStore, applyMiddleware, Store } from 'redux';
import thunk, { ThunkMiddleware } from 'redux-thunk';

import { Reducers, StoreState } from './reducers';
import { UserActions } from './actions';
import axios, { AxiosInstance } from 'axios';

const api = axios.create({
  baseURL: '/api'
});

export function getStore(
  initialState: StoreState
): Store<StoreState, UserActions> {
  return createStore(
    Reducers,
    initialState,
    applyMiddleware<ThunkMiddleware<StoreState, UserActions, AxiosInstance>>(
      thunk.withExtraArgument(api)
    )
  );
}
