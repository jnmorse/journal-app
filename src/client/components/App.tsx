import React, { Component } from 'react';
import { Route, Switch, RouteComponentProps } from 'react-router-dom';

import Home from '../pages/Home';
import { NotFound } from '../pages/NotFound';
import { ConnectedSignup } from '../pages/Signup';
import { ConnectedSignin } from '../pages/Signin';

import JournalForm from '../pages/JournalForm';
import requireAuth from '../hoc/RequireAuth';
import ViewEntry from '../pages/ViewEntry';

export class App extends Component {
  public render(): JSX.Element {
    return (
      <Switch>
        <Route path="/" exact component={Home} />
        <Route path="/signup" component={ConnectedSignup} />
        <Route path="/signin" component={ConnectedSignin} />
        <Route path="/journal/new" component={requireAuth('/')(JournalForm)} />
        <Route exact path="/journal/:id" component={ViewEntry} />
        <Route
          path="/journal/:id/edit"
          component={requireAuth('/')(JournalForm)}
        />
        <Route status={404} component={NotFound} />
      </Switch>
    );
  }
}
