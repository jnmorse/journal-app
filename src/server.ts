import http, { Server } from 'http';
import mongoose from 'mongoose';

import 'dotenv/config';

import { app } from './app';

const PORT: number = Number(process.env.PORT) || 3010;

const httpServer: Server = http.createServer(app);

const mongoUrl = process.env.MONGODB_URL;

if (!mongoUrl) {
  throw new Error('env MONGODB_URL is required to run');
}

mongoose
  .connect(mongoUrl, { useNewUrlParser: true, useCreateIndex: true })
  .then(() => {
    httpServer.listen(PORT, '0.0.0.0', () => {
      // tslint:disable-next-line: no-console
      console.log(`App Running: http://localhost:${PORT}/`);
    });
  })
  // tslint:disable-next-line: no-console
  .catch(error => console.error(error));
