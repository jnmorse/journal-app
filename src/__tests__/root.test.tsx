import mongoose from 'mongoose';
import React from 'react';
import { MongoMemoryServer } from 'mongodb-memory-server';
import request from 'supertest';
import { renderToString } from 'react-dom/server';
import { StaticRouter as Router } from 'react-router-dom';
import { minify } from 'html-minifier';
import { Provider } from 'react-redux';

import { app } from '../app';
import { App } from '../client/components/App';
import { getStore } from '../client/store';

let mongoServer: MongoMemoryServer;

describe('server tests', () => {
  beforeAll(async () => {
    mongoServer = new MongoMemoryServer();
    const mongoURI: string = await mongoServer.getConnectionString();
    await mongoose.connect(mongoURI, {
      useNewUrlParser: true,
      useCreateIndex: true
    });
  });

  afterAll(async () => {
    await mongoose.disconnect();
    await mongoServer.stop();
  });

  test('route: /', async () => {
    const context = {};

    const response = await request(app).get('/');
    const vendorScriptTag: RegExp = /<script src="\/?js\/vendor\.[a-z0-9]+\.js"><\/script>/gu;
    const mainScriptTag: RegExp = /<script src="\/?js\/main\.[a-z0-9]+\.js"><\/script>/gu;

    const appString: string = renderToString(
      <Provider
        store={getStore({
          user: { id: '', email: '', created: '', updated: '' },
          journals: []
        })}
      >
        <Router location="/" context={context}>
          <App />
        </Router>
      </Provider>
    );

    // the status should be 200
    expect(response.status).toBe(200);
    expect(context).toEqual({});

    // Check that text we expect is returned
    expect(response.text).toEqual(expect.stringContaining('DOCTYPE'));
    expect(response.text).toEqual(expect.stringMatching(vendorScriptTag));
    expect(response.text).toEqual(expect.stringMatching(mainScriptTag));
    expect(response.text).toMatchSnapshot();
  });
});
