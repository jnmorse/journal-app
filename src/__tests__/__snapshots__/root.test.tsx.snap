// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server tests route: / 1`] = `"<!DOCTYPE html><html><head><meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\"><meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\"><link rel=\\"stylesheet\\" href=\\"/css/vendor.239d9b53.css\\"><link rel=\\"stylesheet\\" href=\\"/css/main.239d9b53.css\\"><title data-react-helmet=\\"true\\"></title></head><body><div id=\\"root\\"><nav class=\\"navbar navbar-expand-sm navbar-dark bg-dark\\"><header class=\\"sr-only\\"><h2>Site Navigation</h2></header><a href=\\"/\\" class=\\"navbar-brand\\"><img src=\\"\\" alt=\\"Quill\\" height=\\"30\\" style=\\"margin-right:10px\\">Digital Journal</a><button aria-controls=\\"basic-navbar-nav\\" type=\\"button\\" aria-label=\\"Toggle navigation\\" class=\\"navbar-toggler\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"navbar-collapse collapse\\" id=\\"basic-navbar-nav\\"><div class=\\"ml-auto navbar-nav\\"><a href=\\"/posts\\" role=\\"tab\\" tabindex=\\"-1\\" class=\\"nav-link\\">Posts</a><a href=\\"/signin\\" role=\\"tab\\" tabindex=\\"-1\\" aria-selected=\\"false\\" class=\\"nav-link\\">Signin</a><a href=\\"/signup\\" role=\\"tab\\" tabindex=\\"-1\\" aria-selected=\\"false\\" class=\\"nav-link\\">Signup</a></div></div></nav><main><header style=\\"text-align:center\\" class=\\"jumbotron jumbotron-fluid\\"><img src=\\"\\" alt=\\"image\\" width=\\"200\\" style=\\"margin:10px\\"><h2>Digital Journal</h2><p>Organize all your content and ideas together in one place.</p></header><div class=\\"row\\"></div></main><footer>Â© 2019; Joseph Morse; MIT</footer></div><script>window.initState = {\\"user\\":false,\\"journals\\":[]}</script><script src=\\"/js/vendor.239d9b53.js\\"></script><script src=\\"/js/main.239d9b53.js\\"></script></body></html>"`;
